update mW calculation in the build in Standard Model

wkotlarski opened the pull request:
@Expander, I need to update the W-boson mass calculation in the built-in Standard Model. Without this I cannot finish #446.

Expander commented:
I need a few days to review this PR. One thing I need to consider is how this affects Thomas' FlexibleEFTHiggs calculation.

wkotlarski commented:
Sure. This is a work in progress so there's not much for you to review yet. Rather I need some  pointers (put xkcd joke here) how to proceed.

Expander commented:
Just for my understanding: Are you "updating" the calculation in the SM class to the more precise one with the decoupling property? If so, I'd need to keep the old calculation somewhere for the routines from Thomas (maybe in a kind of legacy function/class or so).

Expander commented:
Of course, this makes sense.

wkotlarski commented:
@Expander, I think I've understood the setup for our $m_W$ calculation. Correct me if I'm wrong - I think we:

1. compute $m_W$
2. override the value in `QedQcd` by the one we've computed
3. override the one in the spectrum by the one from `QedQcd`

This is also what the built in SM is trying to do. In `standard_model.cpp`, in `initialise_from_input` there are those lines
```cpp
if (get_thresholds() && threshold_corrections.sin_theta_w > 0)
   qedqcd.setPoleMW(mw_pole);
```
This is a copy of what is also present in autogenerated models. But in those models `qedqcd` is global. Here it's local so this override does nothing or almost nothing. Maybe we should update this to match what is done in autogenerated models.

Second question, I'm confused by all the builtin SM versions we have. In `FlexibleDecay` you've used `standard_model::Standard_model`. So this is also what I tried to follow. But it seems that there's also a two scale model there. Which one should I use in my `HiggsTools` interface?

Expander commented:
> This is a copy of what is also present in autogenerated models. But in those models `qedqcd` is global. Here it's local so this override does nothing or almost nothing. Maybe we should update this to match what is done in autogenerated models.

I see what you mean. The function `initialise_from_input` is intended to fix the running parameters of the SM, this is why it is working on a local copy of `qedqcd`. I.e. it should not overwrite the passed object.

If you are interested in the SM-predicted `M_W`, I think the function `Standard_model::calculate_theta_w` may be a starting point to look at. It returns a pair containing `\theta_W` and the predicted `M_W`. The routines to predict `M_W` are in `weinberg_angle.{hpp,cpp}`. These are quite old routines and maybe they should be updated/changed/adapted. (They are taken from `SOFTSUSY` and contain some MSSM contributions as well, which are not used.)

Expander commented:
> Second question, I'm confused by all the builtin SM versions we have. In `FlexibleDecay` you've used `standard_model::Standard_model`. So this is also what I tried to follow. But it seems that there's also a two scale model there. Which one should I use in my `HiggsTools` interface?

You are right, but this two-scale model is just a wrapper. This wrapper structure is similar to the auto-generated code (sorry for the confusing naming):
~~~
// auto-generated:
<model>_two_scale_model  <-  <model>_mass_eigenstates

// pre-generated SM:
standard_model_two_scale_model  <-  Standard_model
~~~
This two-scale model is intended for scenarios where the SM shall be used with at two-scale solver. There may be another wrapper soon for the shooting solver etc.

Anyway, if you only need the SM parameters, masses and mixing matrices, using `Standard_model` is perfect. :)

Expander commented:
Would you mind briefly sketching how you'd like to use the so-modified `Standard_model` class? I have in principle nothing against these changes, I'm just worried that we're adding some extra state to the `Standard_model` class which may make its usage harder. (But maybe my concerns are pointless and it makes using the class easier.)

wkotlarski commented:
The end goal is that I want `get_physical()` to return the $m_W$ computed in the `weinberg` class, not the one from the standard one-loop calculation. This is also what all autogenerated models do.

Expander commented:
I see. Will you create a PR where we could discuss the details?

wkotlarski commented:
Is there any fundamental difference (apart from how we return results) between `standard_model` `Weinberg_angle::calculate` and `<model>_weiberg_angle::calculate`?

Expander commented:
The former originates from SOFTSUSY that is applicable to the SM and the MSSM. (It contains MSSM contributions, which are not used and could be removed.)

The latter is a generalization to arbitrary models.

But I'm not sure if the former is up-to date with the latter one w.r.t. the SM calculation.

wkotlarski commented:
I'm asking because now on this branch I have 2: which I've cleverly called `calculate` and `calculate2`, where `calculate2` is a copy from the autogenerated `SM`. If possible, I'd like to replace the old `calculate` with this new one.

Expander commented:
Replacing it would indeed be best in my opinion.

wkotlarski commented:
I've simplified the code and removed most of the now unused stuff. The pole $m_W$ returned by `standard_model` now matches exactly the one from the fit formula (see `test/test_standard_model_mw_calculation.cpp`). I still need to fix some of the test I've broken. I know it's a dangerous PR so we can take our time to look at it.

wkotlarski commented:
What renormalization scale is used in `SM_weinberg_angle.cpp`? In `calculate_delta_r_hat` we write
```cpp
      if (Abs(1. - model->get_scale() / mz) > 0.1) {
#if defined(ENABLE_VERBOSE) || defined(ENABLE_DEBUG)
         WARNING("scale deviates from value mz_pole assumed in deltaRHat2LoopSM");
#endif
```
which suggests that $\mu_R = M_Z$. But $\alpha_s$ used in `SM_weinberg_angle` is set in `SM_two_scale_low_scale_constraint.cpp` as
```
calculate_alpha_s_SM5_at(qedqcd, qedqcd.displayPoleMt())
```
So we're using $\alpha_s (M_t)$.

Expander commented:
Thanks for the proposal!

One question/point for discussion (which is related to your first question): There may be another option to implement this, which would avoid storing the `QedQcd` object in the SM class: One could calculate $M_W$ and $\theta_W$ directly in the `initialise_from_input` function and store it in the SM class. The function `calculate_theta_w` would then just return the stored value. What do you think about this?

While thinking about this, my basic worry is the following: `QedQcd` is a user input. If we store it internally in several places and create fancy algorithms to solve BVPs, chances are high that the user input and the internally stored objects become out of sync and cause inconsistencies.

wkotlarski commented:
The reason why I didn't do it like you suggest is that in the SM `calculate_mw_pole_SM_fit_MSbar` depends on one runtime parameter: Higgs pole mass. So ideally, the final $m_W$ should be computed after the `calculate_Mhh_pole` call. Which is not done by `initialise_from_input`.

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -625,9 +625,14 @@ void Standard_model::calculate_pole_masses()
-   calculate_MVWp_pole();
 
 #endif
+   double mt_pole = qedqcd.displayPoleMt();
```
If you don't mind, please move these lines into a separate function, maybe called `calculate_MVWp_pole_fit`.

wkotlarski replied:
Done in 3923909b5e2ead6a8a570a94fd6bc8b94f6d62b6.

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -941,91 +941,55 @@ double Standard_model::calculate_delta_alpha_s(double alphaS) const
+   const double mh_pole = get_mh_pole();
+
+   const auto sm_mw = flexiblesusy::sm_mw::calculate_mw_pole_SM_fit_MSbar(
+      this->get_physical().Mhh, mt_pole, alpha_s_mz, dalpha_s_5_had);
```
Do you want to use `mh_pole` here instead of `this->get_physical().Mhh`?

wkotlarski replied:
Done in e60c13ec590901089b7d59418cf5105a6c3c666d.

Expander replied:
Great, thanks!

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -941,91 +941,55 @@ double Standard_model::calculate_delta_alpha_s(double alphaS) const
-      problems.flag_no_sinThetaW_convergence();
-   else
-      problems.unflag_no_sinThetaW_convergence();
+   const double mt_pole = qedqcd.displayPoleMt();
```
It seems to me that this code snippet is mostly a copy from `calculate_pole_masses`. I think it would be good to re-use the function `calculate_MVWp_pole_fit` from there. (It should get the `mh` and `mw` as parameters, I guess.)

wkotlarski replied:
Maybe solved by 2e3adace5.

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -941,91 +941,55 @@ double Standard_model::calculate_delta_alpha_s(double alphaS) const
+   weinberg.set_number_of_loops(this->get_threshold_corrections().sin_theta_w);
+   weinberg.set_number_of_iterations(number_of_iterations);
+
+   double theta_w;
```
If an exception is thrown, then this variable contains garbage. Please initialize it so some reasonable value, maybe to the old value or the one from `Electroweak_constants`?

wkotlarski replied:
Done in 97eaea3f50f6868be37489878f28d48f1e509a90.

Expander commented on model_specific/SM/standard_model.hpp:
```diff
@@ -657,6 +654,7 @@ class Standard_model : public Beta_function {
    Eigen::Array<double,3,1> MFu{Eigen::Array<double,3,1>::Zero()};
    Eigen::Array<double,3,1> MFe{Eigen::Array<double,3,1>::Zero()};
    double MVWp{};
```
Please add a comment: `// running W boson mass`

wkotlarski replied:
Done in 673aff15e9eb625a99aabbc1f0092c29ab66a93e.

Expander commented on model_specific/SM/standard_model.hpp:
```diff
@@ -657,6 +654,7 @@ class Standard_model : public Beta_function {
    Eigen::Array<double,3,1> MFu{Eigen::Array<double,3,1>::Zero()};
    Eigen::Array<double,3,1> MFe{Eigen::Array<double,3,1>::Zero()};
    double MVWp{};
+   double MVWp_pole{};
```
Please add a comment: `// W boson pole mass (determined from fit formula)` or similar, to avoid confusion.

wkotlarski replied:
Done in 42e877323e3f57726a8e46137890990cdddba3f5.

Expander replied:
Sorry, would you mind changing the comment to
~~~.cpp
double MVWp_pole{}; ///< W boson pole mass (determined from fit formula)
~~~
in order for Doxygen to generate a corresponding entry? Sorry for being pedantic. :)

Expander replied:
(Also for `MVWp` above.)

Expander commented on model_specific/SM/weinberg_angle.cpp:
```diff
@@ -103,20 +66,20 @@ Weinberg_angle::Weinberg_angle()
+Weinberg_angle::Weinberg_angle(
+   const standard_model::Standard_model* model_,
+   const Sm_parameters& sm_parameters_)
+   : number_of_iterations(20)
```
I think it would now be better to initialize these parameters (`number_of_iterations`, ..., `sin_theta`) in the `.hpp` file in order to avoid a redundant initialization in both constructors.

wkotlarski replied:
Done in 55cfdda36c6fbec9c39c35c8b77385d59485f84e.

Expander commented on model_specific/SM/weinberg_angle.cpp:
```diff
@@ -103,20 +66,20 @@ Weinberg_angle::Weinberg_angle()
-
-void Weinberg_angle::disable_susy_contributions()
+Weinberg_angle::Weinberg_angle(
+   const standard_model::Standard_model* model_,
```
I think we should throw an exception of `model_ == nullptr`.

Btw., why are you passing a pointer? Do you want to express that this parameter is somehow optional (if `model_ == nullptr`)? If not, then a reference would be better, I think.

wkotlarski replied:
I need to think about this. I forgot how all of this was supposed to work. Naively, I'd say that it was simply supposed to be a copy of a constructor from `<model>_weinber_angle`, e.g.
```cpp
CLASSNAME::SM_weinberg_angle(
   const SM_mass_eigenstates* model_,
   const Sm_parameters& sm_parameters_)
   : model(model_)
   , sm_parameters(sm_parameters_)
{
}
```

Expander replied:
If there is no reason that the `model_` pointer can be `nullptr`, then I'd prefer to pass `model_` by const reference, not by pointer.

wkotlarski commented:
> What renormalization scale is used in `SM_weinberg_angle.cpp`? In `calculate_delta_r_hat` we write
> 
> ```c++
>       if (Abs(1. - model->get_scale() / mz) > 0.1) {
> #if defined(ENABLE_VERBOSE) || defined(ENABLE_DEBUG)
>          WARNING("scale deviates from value mz_pole assumed in deltaRHat2LoopSM");
> #endif
> ```
> 
> which suggests that μR=MZ. But αs used in `SM_weinberg_angle` is set in `SM_two_scale_low_scale_constraint.cpp` as
> 
> ```
> calculate_alpha_s_SM5_at(qedqcd, qedqcd.displayPoleMt())
> ```
> 
> So we're using αs(Mt).

@Expander, please also remember about this comment. It does bother me.

Expander commented:
I've looked into this: The coupling $\alpha_s(M_t)$ is used only in the 2-loop contributions to $\Delta\hat r$ and $\Delta\hat\rho$, nowhere else. (The 2-loop expressions for $\Delta\hat r$ and $\Delta\hat\rho$ originate from the term $\sim\alpha_s(M_t)$ in Eq.(15d) from https://arxiv.org/pdf/hep-ph/9212285.pdf .)

All the other parts of the calculation of $\theta_W$ are expressed in terms running couplings at $Q=M_Z$, as done in PBMZ (Appendix C) and as written in our MW paper.

wkotlarski commented:
Thanks a lot for having a look at that. I need to now fix tests. There's a lot of tests using the old `Data` structure. I'm now busy with other stuff so I don't know when I will have time to deal with it. But this PR is not very urgent, so we can wait.

Expander commented:
Thanks for this! There is no time pressure on my side either.

wkotlarski commented:
@Expander, are tests in `test_SM_weinberg_angle.cpp` still needed in the current form? They check hard coded expressions in `test_SM_weinberg_angle.cpp` against identical hard coded expressions in `model_specific/SM/weinberg_angle.cpp`. For example, the function `calculate_delta_vb_sm_1loop` in  `test_SM_weinberg_angle.cpp` is a copy of `Weinberg_angle::calculate_delta_vb_sm` in `model_specific/SM/weinberg_angle.cpp`. Maybe it would be enough to compare against hard coded numerical values.

Expander commented:
> @Expander, are tests in `test_SM_weinberg_angle.cpp` still needed in the current form? They check hard coded expressions in `test_SM_weinberg_angle.cpp` against identical hard coded expressions in `model_specific/SM/weinberg_angle.cpp`. For example, the function `calculate_delta_vb_sm_1loop` in `test_SM_weinberg_angle.cpp` is a copy of `Weinberg_angle::calculate_delta_vb_sm` in `model_specific/SM/weinberg_angle.cpp`. Maybe it would be enough to compare against hard coded numerical values.

You are right, comparing `1 == 1` is not very useful. A hard-coded number (with some reasonable uncertainty) may be sufficient.

wkotlarski commented:
About overriding the $m_W$ computed from 2-point function with the one from muon decay: from the point of view of decay calculation it is enough if `QedQCD` object returns correct $m_W$ (i.e. calculated via muon decay). So there 2 options:

1. override $m_W$ in QedQCD object with the one from muon decay
2. and/or override the one returned by `get_physical()` with the one from muon decay

Both autogenerated and build in models should always behave the same in this respect and either of the above options would work for me (as far as decays are concerned). Which option would be best in your opinion? 1, 2 or 1 and 2?

Expander commented:
Hmm, I think it would be best if we'd keep the convention that a `QedQcd` contains only SM quantities. So, if the W pole mass `M_W` that you need contains BSM contributions, it would be best to store it in `get_physical()`. This also makes sense because the `Physical` struct belongs to a BSM model and is supposed to contain their pole masses.

wkotlarski commented:
Hey @Expander. I've fixed most of tests. The setup is now as follows: `calculate_threshold_corrections()` sets physical `mW`. Then, `calculate_pole_masses()` does the following check
```cpp
if (!(PHYSICAL(MVWp) > 0.)) {
   calculate_MVWp_pole();
}
```
This is needed since in some tests we setup models without doing the matching (that is without calling `calculate_threshold_corrections()`). The above check ensures that `mW` is not 0.

The remaining issue now is what to do with `test_CMSSM_weinberg_angle_meta` and `test_CMSSM_weinberg_angle`. The `Weinber_angle` class in the  builin `standard_model` now does what it says on the tin. It's only a SM calculation. Before it was a copy of softsusy routines, including susy contributions. The above `CMSSM` test were checking against them. What to do with those tests?

Expander commented on meta/LoopMasses.m:
```diff
@@ -1031,22 +1031,33 @@ be set to the (positive) tree-level mass.  M_tree
+                          "tp.run_task([this] () {\n" <>
+                          IndentText[
+                             "if (!(PHYSICAL(M" <> ToString@GetWBoson[] <> ") > 0.)) {\n" <>
+                                IndentText["calculate_M" <> ToString@GetWBoson[] <> "_pole();\n"] <>
```
Please use ``LoopMasses`CreateLoopMassFunctionName[GetWBoson[]]``, which generates `"calculate_MWVm_pole"`.

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -4900,6 +4858,24 @@ std::ostream& operator<<(std::ostream& ostr, const Standard_model& model)
+   return qedqcd_tmp.displayAlpha(softsusy::ALPHAS);
+}
+
+void Standard_model::calculate_MVWp_pole_fit(const double mh_pole) {
```
Please replace `const double mh_pole` by `double mh_pole` in the function argument list.

wkotlarski replied:
Ok. Side note: I actually like this syntax. Even though simple types are passed via copy, making them const does not allow you to change the value of the copy. So you know that if you use this variable, it's exactly what was passed to the function. Which is nice for very long functions that may try to do something with the passed argument before you come to using it.

Expander replied:
Hmm, that's a valid point. Ok, then we could leave it as it is, if you prefer. :)

Expander commented:
I'm fine with these changes.

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -613,10 +614,19 @@ void Standard_model::calculate_pole_masses()
+   tp.run_task([this] () {
+      // if mW=0 it means that the SM was not initialized via matching to the SM
+      // compute mW using ordinary 1-loop calculation
+      if (!(PHYSICAL(MVWp) > 0.)) {
```
I'm wondering if the test should be `if (PHYSICAL(MVWp) == 0.)`.
Reason: The current condition would be true if the mass is `<= 0`, i.e. a tachyonic mass would also trigger the re-calculation of the W boson mass.

wkotlarski replied:
So maybe `if (!(std::abs(PHYSICAL(MVWp)) > 0.))`? I wrote it like that because I'm always afraid of comparing a double with 0.

Expander replied:
I know what you mean. However, if we have initialized `MVWp = 0`, then the statement `PHYSICAL(MVWp) == 0.` is true if `MVWp` was not touched (computed). So, this is safe. :)

Expander commented on model_specific/SM/standard_model.cpp:
```diff
@@ -625,8 +635,12 @@ void Standard_model::calculate_pole_masses()
    calculate_MFe_pole();
-   calculate_MVWp_pole();
-
+   if (!(PHYSICAL(MVWp) > 0.)) {
```
see my comment above

wkotlarski commented:
I see that some tests in BSM models have shifted. This is cause by a change of this line
```
qedqcd.setPoleMW(result.second);
```
to this
```
Pole(M" <> ToString@TreeMasses`GetWBoson[] <> ") = result.second;
```
in `meta/ThresholdCorrections.m`. The previous setup was actually weird. It worked like this

1. `calculate_theta_w()` initalizes `Sm_parameters` with `mW` from `qedqcd`
2. we compute BSM `mW` using `qedqcd` value as SM reference number and store it in `qedqcd`

I think that the `qedqcd` object is persistent between iterations. Doesn't it mean that on the next iteration, we again compute BSM corrections, but this time to a BSM `mW` mass?

Expander commented:
> I think that the `qedqcd` object is persistent between iterations. Doesn't it mean that on the next iteration, we again compute BSM corrections, but this time to a BSM `mW` mass?

The low-scale constraint class has a local copy of `qedqcd` which can be modified. So, the old approach was kind of a fixed-point iteration, where you started with some value of MW, which was fed into the equations, and you get a new value of MW out, which is closer to the true result.

I think in the new approach the calculation of MW only depends on the fixed SM input quantities (which don't change between iterations) and the Higgs pole mass (which changes between iterations).

wkotlarski commented:
So I'll simply update the tests. Shifts are not large. 1e-5 - 1e-6 for `amm`.

wkotlarski commented:
There's something weird with built in softsusy libraries. Apart form issue #588 (this is a generic problem, unrelated to this PR), the decoupling test `test_MRSSM2_mw_calculation.cpp` now fails (but only with softsusy). With `mh >= 9999.8` we get
```
unknown location(0): fatal error: in "test_decoupling": flexiblesusy::NoConvergenceError: 
test/test_MRSSM2_mw_calculation.cpp(76): last checkpoint: "test_decoupling" test entry
```
This test work fine for `mh = 9999.7` and for `mh = 10000` but with Collier or LoopTools.

I've lowered maximal mass in this test, left an explanation why and tightened the comparison precision (see 314048507f93dc58940e25cf4229b131ba1fa2e4).

Expander commented:
I find this not surprising. For large SUSY scales the high-scale <-> low-scale iteration is very unstable and very sensitive on the details of the loop calculations and loop functions. Since in this PR you've change details of the calculation, such convergence failures may happen. Your fix is fine.

wkotlarski commented:
Great. Once all tests pass, I'll merge it. Thanks a lot for the help.
